README
Task-Kensho-Testing
-------------------

From http://en.wikipedia.org/wiki/Kensho:

    Kenshō (見性) (C. Wu) is a Japanese term for enlightenment experiences—most
    commonly used within the confines of Zen Buddhism—literally meaning
    "seeing one's nature"[1] or "true self."[2] It generally "refers to the
    realization of nonduality of subject and object."[3]

Task::Kensho is a first cut at building a list of recommended modules for
Enlightened Perl development. CPAN is wonderful, but there are too many wheels
and you have to pick and choose amongst the various competing technologies.

The plan is for Task::Kensho to be a rough testing ground for ideas that go
into among other things the Enlightened Perl Organisation Extended Core
(EPO-EC).

The modules that are bundled by Task::Kensho are broken down into several
categories and are still being considered. They are all taken from various top
100 most used perl modules lists and from discussions with various subject
matter experts in the Perl Community. That said this bundle does I<not> follow
the guidelines established for the EPO-EC for peer review via industry
advisers.

Starting in 2011, Task::Kensho split its sub-groups of modules into
individually-installable tasks.  Each Task::Kensho sub-task is listed at the
beginning of its section in this documentation.

When installing Task::Kensho itself, you will be asked to install each
sub-task in turn, or you can install individual tasks separately. These
individual tasks will always install all their modules by default. This
facilitates the ease and simplicity the distribution aims to achieve.


Changes
Revision history for Task-Kensho-Testing

0.38      2014-09-13
          - WebDev: Template::Toolkit is not indexed (see RT#92588) -
            refer to Template instead
          - subtask distributions now just use regular prereqs on their
            component modules, rather than using optional_features metadata
          - Task-Kensho now once again prompts at install time for each
            subtask to be optionally installed

0.37      2014-08-16
          - ModuleDev: Switch Carp::Always over to Devel::Confess (per mst)
          - ModuleDev: Add Data::Printer (per Sawyer X)
          - ModuleDev: Add Devel::Dwarn (per mst)
          - Toolchain: Add Pinto (per Sawyer X)
          - Toolchain: Add App::FatPacker (per perigrin)

0.36      2014-03-08
          - fix pod markup error in main Task::Kensho distribution
          - module list revisions to CLI, Dates, Toolchain, WebCrawling,
            WebDev

0.35      2014-01-30
          - updated recommendations for Async, CLI, ModuleDev, Scalability,
            WebDev

0.34004   2014-01-22 (TRIAL RELEASE)
          - still more pod and metadata tweaks

0.34003   2014-01-22 (TRIAL RELEASE)
          - more pod and metadata tweaks

0.34002   2014-01-21 (TRIAL RELEASE)
          - fix mangled abstract in META.*
          - fixed missing list of recommendations in Task::Kensho pod

0.34001   2014-01-21 (TRIAL RELEASE)
          - fix munging of ::Hackery preamble into stopwords
          - fixed bugtracker metadata (now indicating github issues)
          - distributions split apart again, with new! and enhanced!
            Dist::Zilla generation magic

0.34      2013-08-20 01:27:07Z (TRIAL RELEASE)
          - distributions reunified under Task-Kensho, and now auto-generated
            from a master YAML file using Dist::Zilla

0.33      2012-06-24
          - Task::Kensho::Testing:
            - Retired Test::Exception and replaced with Test::Fatal
            - added Test::Deep, Test::Requires and Test::NoWarnings
          - Task::Kensho::WebDev: added Plack

0.32      2012-06-08
          - Fix reference to Module::CoreList (RT#65948)
          - Add Plack to Task::Kensho::WebDev

0.31      2011-04-11
          - Fix Task::Kensho::Logging version, thanks kent

0.30      2011-04-11
          - new group, Task::Kensho::Logging

0.29      2011-03-25
          - Task::Kensho::CLI: Fix Module::CoreList link in POD, thanks alister@alisterwest.com.
          - Task::Kensho::WebCrawling: Bundle::LWP is gone; replace with
            LWP::UserAgent. Thanks, kentnl@cpan.org.

0.28      2011-01-23
          - Bugfix: Some tasks had a version of 1.0 marked as required. Fix this.
          - Task::Kensho::CLI: It's App::Nopaste, not App::NoPaste.

0.27      2011-01-20
          - Split Task::Kensho into sub-tasks which all install their modules
            by default. Task::Kensho itself will prompt the user to install
            these subtasks. Thanks to perigrin for the idea.

0.26      2010-10-24 # Late ship date due to Task::Catalyst issue with BerkeleyDB
          - Identical to 0.25

0.25      2010-09-19
          - Add Email::Simple
          - Add Email::MIME::Kit

0.24      2010-08-26 # Now with iCal reminders so I don't slip a date
          - Add autodie
          - Actually install App::Nopaste
          - Make sure the POD reflects the Makefile.PL.

0.23      2010-07-19
          - Add CHI under Scalability

0.22      2010-04-19
          - Replace Bundle::CPAN with App::cpanminus
          - Added HTTP::Lite which is used to great effect by both App::cpanminus and App::perlbrew
          - Added App::perlbrew
          - Totally missed March

0.21      2010-02-21
          - Add XML::LibXSLT (per mst)
          - Add WWW::Mechanize::TreeBuilder (per mst)
          - Add CPAN::Uploader (per mst)
          - Make Template::Toolkit an explicit dependency (per mst)

0.20      2010-01-20
          - Change to depend on Task::Catalyst rather than explicit
            dependencies.
          - Remove CGI::FormBuilder::Source::Perl, this should probably be
            handled by Task::Catalyst
          - Add XML::Atom
          - Add Docs for XML::RSS
          - Change to depend on Task::Moose rather than explicit dependencies.
          - Make Task::Moose a default

0.19      2009-12-28 #        Of course I blow my first release date.
          - Remove MooseX::AttributeHelpers
          - Add MooseX::Aliases
          - Document the move to Monthly Releases for Task::Kensho
          - Remove reference to gitorious
          - Add Perl::Critic as suggested by Hinrik via rt#48295
          - Add Devel::Cover as suggested by Hinrik via rt#48293
          - Add Time::Tiny and DateTime::Tiny as suggested by mst

0.18      2009-11-19
          - Replace Module::Starter with Dist::Zilla
          - Remove Time::y2038
          - Add Try::Tiny and TryCatch

0.17      2009-08-28
          - Fix broken Pod::Readme link in the Makefile.PL

0.16      2009-08-28
          - Remove XXX
          - Add POD::Readme and Software::License at the request of CSJewell
            so that we're inline with the upcoming Perl::Dist::Satori

0.15      2009-07-26
          - Comment out XXX since it can't be found by CPAN anymore.

0.14      2009-07-16
          - Add XML modules
          - Update Documentation a bit to reflect Task::Kensho's place in the
            universe better

0.13      2009-07-02
          - Identical to 0.12 - fix MANIFEST.SKIP and re-release

0.12      2009-06-23 (Failed release)
          - Time::y2038
          - Perl::Version

0.11      2009-05-25
          - Change from namespace::clean to namespace::autoclean (rafl)

0.10      2009-05-23
          - Change dep for XXX to YYY.pm as per #epo

0.0.7     2008-12-12
          - Add MooseX::StrictConstructor as per Dave Rolsky and Ricardo Signes
          - Add MooseX::SimpleConfig

0.0.6     2008-12-12
          - Fix Module::CoreList to be spelled with a capital L


Build.PL
# This Build.PL for Task-Kensho-Testing was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.006
use strict;
use warnings;

if (eval 'use Module::Build::Tiny 0.038; 1')
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";

    # This section for Task-Kensho-Testing was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.007.
    use strict;
    use warnings;

    use 5.006;
    # use Module::Build::Tiny 0.038;
    Build_PL();
}
else
{
    $ENV{PERL_MB_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.

EOW
    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "build_requires" => {},
      "configure_requires" => {
        "ExtUtils::MakeMaker" => 0,
        "Module::Build::Tiny" => "0.038"
      },
      "dist_abstract" => "A Glimpse at an Enlightened Perl: Testing",
      "dist_author" => [
        "Chris Prather <chris\@prather.org>"
      ],
      "dist_name" => "Task-Kensho-Testing",
      "dist_version" => "0.38",
      "license" => "perl",
      "module_name" => "Task::Kensho::Testing",
      "recommends" => {},
      "recursive_test_files" => 1,
      "requires" => {
        "Devel::Cover" => 0,
        "Test::Deep" => 0,
        "Test::Fatal" => 0,
        "Test::Memory::Cycle" => 0,
        "Test::Most" => 0,
        "Test::NoWarnings" => 0,
        "Test::Pod" => 0,
        "Test::Pod::Coverage" => 0,
        "Test::Requires" => 0,
        "Test::Simple" => 0,
        "perl" => "5.006",
        "strict" => 0,
        "warnings" => 0
      },
      "script_files" => [],
      "test_requires" => {
        "ExtUtils::MakeMaker" => 0,
        "File::Spec" => 0,
        "Test::More" => 0
      }
    );


    my %fallback_build_requires = (
      "ExtUtils::MakeMaker" => 0,
      "File::Spec" => 0,
      "Test::More" => 0
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);

    $build->create_build_script;
}


lib/Task/Kensho/Testing.pm
use strict;
use warnings;
package Task::Kensho::Testing;
# git description: v0.36-32-ge84ae83
$Task::Kensho::Testing::VERSION = '0.37';
# ABSTRACT: A Glimpse at an Enlightened Perl: Testing

__END__

=pod

=encoding UTF-8

=head1 NAME

Task::Kensho::Testing - A Glimpse at an Enlightened Perl: Testing

=head1 VERSION

version 0.37

=head1 SYNOPSIS

    > cpanm --interactive Task::Kensho::Testing

=head1 DESCRIPTION

=for stopwords Buddhism EPO Kenshō nonduality amongst Organisation installable

From L<http://en.wikipedia.org/wiki/Kensho>:

=over 4

Kenshō (見性) (C. Wu) is a Japanese term for enlightenment
experiences - most commonly used within the confines of Zen
Buddhism - literally meaning "seeing one's nature"[1] or "true
self."[2] It generally "refers to the realization of nonduality of
subject and object."[3]

=back

L<Task::Kensho> is a list of recommended modules
for Enlightened Perl development. CPAN is wonderful, but there are too
many wheels and you have to pick and choose amongst the various
competing technologies.

The plan is for L<Task::Kensho> to be a rough testing ground for ideas that
go into among other things the Enlightened Perl Organisation Extended
Core (EPO-EC).

The modules that are bundled by L<Task::Kensho> are broken down into
several categories and are still being considered. They are all taken
from various top 100 most used perl modules lists and from discussions
with various subject matter experts in the Perl Community. That said,
this bundle does I<not> follow the guidelines established for the EPO-EC
for peer review via industry advisers.

Starting in 2011, L<Task::Kensho> split its sub-groups of modules into
individually-installable tasks.  Each L<Task::Kensho> sub-task is listed at the
beginning of its section in this documentation.

When installing L<Task::Kensho> itself, you will be asked to install each
sub-task in turn, or you can install individual tasks separately. These
individual tasks will always install all their modules by default. This
facilitates the ease and simplicity the distribution aims to achieve.

=head1 RECOMMENDED MODULES

=head2 L<Task::Kensho::Testing>: Testing

=head3 L<Devel::Cover>

Code coverage metrics for Perl

=head3 L<Test::Deep>

Test deep data structures

=head3 L<Test::Fatal>

Test exception-based code

=head3 L<Test::Memory::Cycle>

Check for memory leaks and circular memory references

=head3 L<Test::Most>

Most commonly needed test functions and features.

=head3 L<Test::NoWarnings>

Ensure that no (unexpected) warnings occur during testing

=head3 L<Test::Pod>

Check for POD errors in files

=head3 L<Test::Pod::Coverage>

Check for pod coverage in your distribution.

=head3 L<Test::Requires>

Make running a test conditional on a particular module being installed

=head3 L<Test::Simple>

Basic utilities for writing tests.

=head1 INSTALLING

Since version 0.34, L<Task::Kensho> has made use of the C<optional_features> field
in distribution metadata. This allows CPAN clients to interact with you
regarding which modules you wish to install.

The C<cpanm> client requires interactive mode to be enabled for this to work:

    cpanm --interactive Task::Kensho::Testing

=head1 BUGS AND LIMITATIONS

This list is by no means comprehensive of the "Good" Modules on CPAN.
Nor is this necessarily the correct path for all developers. Each of
these modules has a perfectly acceptable replacement that may work
better for you. This is however a path to good perl practice, and a
starting place on the road to Enlightened Perl programming.

Please report any bugs or feature requests to
L<https://github.com/EnlightenedPerlOrganisation/task-kensho/issues>.

=head1 SEE ALSO

L<http://www.enlightenedperl.org/>,
L<Perl::Dist::Strawberry|Perl::Dist::Strawberry>

=head1 AUTHOR

Chris Prather <chris@prather.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by Chris Prather.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 CONTRIBUTORS

=for stopwords Karen Etheridge Leo Lapworth Chris Nehren

=over 4

=item *

Karen Etheridge <ether@cpan.org>

=item *

Leo Lapworth <leo@cuckoo.org>

=item *

Chris Nehren <apeiron@cpan.org>

=back

=cut


Makefile.PL
# This Makefile.PL for Task-Kensho-Testing was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.013
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.25.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
    'ExtUtils::MakeMaker' => '0',
    'Module::Build::Tiny' => '0.038',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    $ENV{PERL_MM_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are doing something overly clever, in which case you
should consider setting the 'prefer_installer' config option in CPAN.pm, or
'prefer_makefile' in CPANPLUS, to 'mb" and '0' respectively.

You can also silence this warning for future installations by setting the
PERL_MM_FALLBACK_SILENCE_WARNING environment variable.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.006;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A Glimpse at an Enlightened Perl: Testing",
  "AUTHOR" => "Chris Prather <chris\@prather.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build::Tiny" => "0.038"
  },
  "DISTNAME" => "Task-Kensho-Testing",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Task::Kensho::Testing",
  "PREREQ_PM" => {
    "Devel::Cover" => 0,
    "Test::Deep" => 0,
    "Test::Fatal" => 0,
    "Test::Memory::Cycle" => 0,
    "Test::Most" => 0,
    "Test::NoWarnings" => 0,
    "Test::Pod" => 0,
    "Test::Pod::Coverage" => 0,
    "Test::Requires" => 0,
    "Test::Simple" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.38",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "Devel::Cover" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::Deep" => 0,
  "Test::Fatal" => 0,
  "Test::Memory::Cycle" => 0,
  "Test::More" => 0,
  "Test::Most" => 0,
  "Test::NoWarnings" => 0,
  "Test::Pod" => 0,
  "Test::Pod::Coverage" => 0,
  "Test::Requires" => 0,
  "Test::Simple" => 0,
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
